### PR TITLE
Global POC review

### PR DESCRIPTION
**Pull request by syalioune**

### CODE REVIEW

File: src/main/java/org/acme/Main.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    public static class MyApp implements QuarkusApplication {
+        @Inject
+        CweDataProducer cweRecordGenerator;
+        Logger logger = Logger.getLogger("poc");
+
+        @Override
+        public int run(String... args) {
+            try {
+                Map<Integer, String> cweList = new CweImporter().processCweDefinitions();
```
Comment (syalioune) [inline src/main/java/org/acme/Main.java:29]: The CWE definition is parsed from a static file that is not subject to dynamic change. I think that there are no reasons to justify putting them into Kafka as a Global KTable. It makes things much more complicated that it needs to be : Each instance of the application is going to parse that file, to produce the same events to Kafka, which are going to be dedup and replicated to the same instances for creating a distributed local copy on disk with the exact same content that the file `cwec_v4.6.xml`.

Using the static file is simpler and still consistent with the usage of Apache Kafka.

File: src/main/java/org/acme/consumer/CacheReader.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @ConfigProperty(name = "cache.global.ktable.topic")
+    String cacheGKTable;
+
+    @ConfigProperty(name = "component.cache.global.ktable.store.name")
+    String cacheStoreName;
+    void onStart(@Observes StartupEvent event) {
+        Properties props = new Properties();
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, cacheTopic);
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, server);
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,"latest");
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/CacheReader.java:53]: Using `latest` feels a bit dangerous here as we might miss some events if the consumer is off for a sufficient amount of time to have the offsets purged from `__consumer_offsets` topic.
I would personally go for `earliest` and use the property `consumer.offset`.

File: src/main/java/org/acme/consumer/CacheReader.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        StreamsBuilder builder = new StreamsBuilder();
+        ObjectMapperSerde<ComponentAnalysisCache> componentSerde = new ObjectMapperSerde<>(ComponentAnalysisCache.class);
+        ObjectMapperSerde<CacheKey> cacheSerde = new ObjectMapperSerde<>(CacheKey.class);
+        GlobalKTable<CacheKey, ComponentAnalysisCache> componentCache = builder.globalTable(cacheTopic, Materialized.<CacheKey, ComponentAnalysisCache, KeyValueStore<Bytes, byte[]>>as(cacheStoreName)
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/CacheReader.java:57]: Building a `Streams` topology and a global `KTable` just to have a state store to query feels a bit off. Even more so because the KTable is not used in a Streams Topology.

In order to be more idiomatic with Kafka usage, I will try to leverage directly this `KTable` in the `PrimaryConsumer` stream topology to perform a KStream - KTable join instead of manually performing the join in `BaseComponentAnalyzer.isCacheCurrent`. 

A `KStream - KTable` being windowed, it gives you the opportunity to let's say buffer N events on the same component for `T` seconds in order to let some earlier events to perform the call the OSSIndex / Snyk before populating the vulnerability cache and trigger the join for the N events.

This way, you can avoid invoking the analyzer tasks if it is not necessary and make better use of Apache Kafka.

This would off course need to be described and design more formally.

File: src/main/java/org/acme/consumer/ConfigConsumer.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import java.util.Properties;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Objects;
+
+@ApplicationScoped
+public class ConfigConsumer {
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/ConfigConsumer.java:30]: The use case of putting the configuration properties into Apache Kafka and reading them through a `GlobalKTable` seems overly complicated to me.

1. Configuration need to have a single source of truth to avoid complex debugging. Replicating it into Kafka which make it so that at any point in time a given property could have different values in `API Server database`, in `Kafka` and in each of the consumer instances. Even though, a global KTable, a local state is built on each instance from the Kafka topic so inconsistencies is not out of the way.
2. If we think about the next designs steps, a config property like `snyk.api.token` would be directly owned by the current module and not `API server` so it make sense to not have to go through Kafka to fetch the value

I would personally directly fetch the value from the source by either accepting to :

1. Read them directly from `API server` database with the assumption that at some point that module is going to have a separate database with its own config properties
2. Read them directly from `API server` using REST API providing that some cache capability is taken into account

File: src/main/java/org/acme/consumer/CweConsumer.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.QueryableStoreTypes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import java.util.Objects;
+import java.util.Properties;
+
+@ApplicationScoped
+public class CweConsumer {
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/CweConsumer.java:29]: If you follow my line of thought in `Main` review, that consumer is not required.

File: src/main/java/org/acme/consumer/OSSIndexBatcher.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Properties;
+import java.util.concurrent.LinkedBlockingQueue;
+
+@ApplicationScoped
+public class OSSIndexBatcher {
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/OSSIndexBatcher.java:33]: This stream topology could be merged with the main in `PrimaryConsumer` and is much more complicated than it should be.
I feel like what you're doing here could be resolved with out of the box with [Windowed Aggregation](https://docs.confluent.io/platform/current/streams/developer-guide/dsl-api.html#aggregating) instead of creating and maintaining a processor with custom defined punctuator.

File: src/main/java/org/acme/consumer/PrimaryConsumer.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, server);
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, offset);
+
+
+        StreamsBuilder builder = new StreamsBuilder();
+        ObjectMapperSerde<VulnerabilityAnalysisEvent> vulnerabilityAnalysisEventSerde = new ObjectMapperSerde<>(
+                VulnerabilityAnalysisEvent.class);
+        ObjectMapperSerde<Component> componentSerde = new ObjectMapperSerde<>(Component.class);
+
+        KStream<String, VulnerabilityAnalysisEvent> kStreamsVulnTask = builder.stream(inputTopic, Consumed.with(Serdes.String(),vulnerabilityAnalysisEventSerde )); //receiving the message on topic event
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/PrimaryConsumer.java:61]: We have to be careful with the `inputTopic` here. Based on the structure, I infer that it will be some kind of event generated after `BomUploadEvent` in `API_SERVER`.

Keep in mind the maximum message size for Kafka is `1MB`. Although, it could be increased, this is not recommended as many default settings relies on that assumption. For project with large number of component, this could cause issues.

My recommendation would be to have :

1. An event with the project UUID as reference
2. The consumer use the project UUID to download the BOM using the REST API and produces the relevant just as it could have been done by `flatMapValues` below

File: src/main/java/org/acme/consumer/PrimaryConsumer.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+        //Splitting the incoming message into number of components
+        //Setting the message key same as the name of component to send messages on different partitions of topic event-out
+        KStream<String, Component> splittedStreams = kStreamsVulnTask.flatMapValues(value -> value.getComponents()).selectKey((key, value) -> value.getName());
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/PrimaryConsumer.java:65]: FYI, it would have been simpler to just use flatMap which allow key and value mapping in one go :wink:

File: src/main/java/org/acme/consumer/PrimaryConsumer.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+       /* if (isConsumerEnabled(ConfigPropertyConstants.SCANNER_OSSINDEX_ENABLED.getPropertyName())) {
+            splittedStreams.to(ossTopic, (Produced<String, Component>) Produced.with(Serdes.String(), componentSerde));
+        }
+        if (isConsumerEnabled(ConfigPropertyConstants.SCANNER_SNYK_ENABLED.getPropertyName())) {
+            splittedStreams.to(snykTopic, (Produced<String, Component>) Produced.with(Serdes.String(), componentSerde));
+        }*/
+        splittedStreams.to(ossTopic, (Produced<String, Component>) Produced.with(Serdes.String(), componentSerde));
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/PrimaryConsumer.java:73]: There is no need to split into two different topics as the events are the same. Using different consumer groups for Snyk/OSSIndex, you can read independently from the same topic.

File: src/main/java/org/acme/consumer/PrimaryConsumer.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        splittedStreams.to(snykTopic, (Produced<String, Component>) Produced.with(Serdes.String(),  componentSerde));
+        streams = new KafkaStreams(builder.build(), props);
+        streams.start();
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/PrimaryConsumer.java:76]: If you follow my recommendations, and produce individual components events from BOM download, I would then create a more complete stream topology with the following steps :
- Create a KStream from `component` topic (a component with CPE and PURL would result in two different events)
- Create a Ktable for `vulnerabilities` topic
- Perform a `KStream - KTable` join on `component` and `vulnerabilities` topic with a time window
- Branch to separate components with match from vulnerability cache (which could be forwarded to result topic) from components which should be forwarded to analyzers and then again branch by CPE and PURL.
- Perform a windowed aggregation before submitting groups to the relevant analyzer using `flatMapValues` to send results downstream to result topic. At that point in time `vulnerabilities` topic should be updated to also update the `KTable`

Note: [There should be some kind of processor to periodically expire data from the vulnerability KTable based on a TTL](https://developer.confluent.io/tutorials/schedule-ktable-ttl-aggregate/kstreams.html)

File: src/main/java/org/acme/consumer/SnykAnalyzer.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        logger.info("Printing topic name here: " + topic);
+        logger.info(payload);
+        logger.info("Printing payload purl here: " + payload.getPurl());
+        logger.info(payload.getName());
+        ArrayList<Component> componentArrayList = new ArrayList<>();
+        componentArrayList.add(component);
+        snykAnalysisEvent.setComponents(componentArrayList);
+        snykTask.inform(snykAnalysisEvent);
+        // ack will commit the latest offsets (per partition) of the batch.
+        return records.ack();
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/SnykAnalyzer.java:47]: With this implementation, there a risk of not processing some events as `SnykAnalysisTask` perform some asynchronous tasks with `SnykAnalysisTaskUtil`. if the Thread encounters some errors whereas the event offset has been commited to Kafka, it won't be replayed and thus will be lost :cry:

c.f Quarkus Kafka documentation

`latest commits the record offset received by the Kafka consumer as soon as the associated message is acknowledged (if the offset is higher than the previously committed offset). This strategy provides at-least-once delivery if the channel processes the message without performing any asynchronous processing. This strategy should not be used in high load environment, as offset commit is expensive. However, it reduces the risk of duplicates.`

File: src/main/java/org/acme/consumer/VulnCacheReader.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.QueryableStoreTypes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import java.util.Objects;
+import java.util.Properties;
+
+@ApplicationScoped
+public class VulnCacheReader {
```
Comment (syalioune) [inline src/main/java/org/acme/consumer/VulnCacheReader.java:30]: Same comment as my previous ones regarding isolated `GlobalKTable`s

File: src/main/java/org/acme/producer/CacheProducer.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    private static final Producer<CacheKey, ComponentAnalysisCache> producer;
+
+
+    static {
+        final var properties = new Properties();
+        properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+        properties.put(ProducerConfig.CLIENT_ID_CONFIG, "CacheProducer");
+        properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, CacheKeySerializer.class.getName());
+        properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ComponentAnalysisCacheSerializer.class.getName());
+        producer = new KafkaProducer<>(properties);
```
Comment (syalioune) [inline src/main/java/org/acme/producer/CacheProducer.java:39]: You should configure the `KafkaProducer` with more resilient settings :

- [Acks](https://docs.confluent.io/platform/current/installation/configuration/producer-configs.html#producerconfigs_acks) to `all`
- [Delivery timeout](https://docs.confluent.io/platform/current/installation/configuration/producer-configs.html#producerconfigs_delivery.timeout.ms) to control retries

I don't think that [max.in.flight.requests.per.connection](https://docs.confluent.io/platform/current/installation/configuration/producer-configs.html#producerconfigs_max.in.flight.requests.per.connection) is relevant here because order is not that much of a requirement here.

File: src/main/java/org/acme/producer/CacheProducer.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+
+    public void sendVulnCacheToKafka(CacheKey key, ComponentAnalysisCache cacheValue) {
+        producer.send(new ProducerRecord<>(cacheTopic, key, cacheValue));
```
Comment (syalioune) [inline src/main/java/org/acme/producer/CacheProducer.java:44]: Careful here, `producer.send` is asynchronous and returns a `Future`. You have performed a `Fire and forget`, there is no guarantee that event has been correctly send. You should provide a `Callback` to assess the result.

**Review by syalioune — COMMENTED** [review]
I think that it is good base to start from. The component could be greatly simplified if Kafka is principally used for what it should be : a communication pipe with some streaming capabilities. Putting to much referential data into Kafka (i.e. State Store) will only add complexity.
There are still several design choices to tackle regarding (some that I haven't consider in this short review) :

- The responsabilities of DT and this module especially regarding which own some business entities (e.g configuration properties)
- The idiomatic use of Apache Kafka
- Accurate configuration of Kafka `consumers`, `producers` and `streams` application
- The processing semantic that we want to achieve (i.e. `at_least_once`, `at_most_once`, `exactly_once`) which impact configuration and the implemntation
- The `poison pills` handling strategy
- The strategy for global periodic portfolio analysis

File: src/main/java/org/acme/Main.java (hunk 1 above)
Comment (mehab) [reply src/main/java/org/acme/Main.java:29]: For this one I have modified the code to not use ktable anymore. Now using a static hashmap that gets loaded on startup in the same way this ktable was getting loaded.

File: src/main/java/org/acme/consumer/ConfigConsumer.java (hunk 4 above)
Comment (mehab) [reply src/main/java/org/acme/consumer/ConfigConsumer.java:30]: For this one we had a slightly modified idea around what you proposed. 
Let’s have a notification topic that sends a notification to our stateless quarkus app whenever there is a change in DT config. If there is a change in config then this notification topic will have an event. We want this event to be listened to by one consumer. When an event occurs, it gets the latest config from DT and updates the info in the in memory structure of quarkus apps. This way we would not be storing global config as such in ktable (like suggested yesterday) and we would only be fetching latest config from dt when there is a change (otherwise use existing values) — so there is no unnecessary polling. I think it is a good idea. What do you think?

File: src/main/java/org/acme/consumer/CacheReader.java (hunk 2 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/CacheReader.java:53]: Thanks for the input Alioune..this is done

File: src/main/java/org/acme/consumer/OSSIndexBatcher.java (hunk 6 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/OSSIndexBatcher.java:33]: Thanks for the input Alioune..this is done

File: src/main/java/org/acme/producer/CacheProducer.java (hunk 14 above)
Comment (japurva1502) [reply src/main/java/org/acme/producer/CacheProducer.java:44]: Thanks for the input Alioune..this is done

File: src/main/java/org/acme/consumer/ConfigConsumer.java (hunk 4 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/ConfigConsumer.java:30]: Implemented this using HashMap instead of Global K Table

File: src/main/java/org/acme/consumer/PrimaryConsumer.java (hunk 10 above)
Comment (mehab) [reply src/main/java/org/acme/consumer/PrimaryConsumer.java:76]: Thanks Alioune.. added this change.

File: src/main/java/org/acme/producer/CacheProducer.java (hunk 13 above)
Comment (mehab) [reply src/main/java/org/acme/producer/CacheProducer.java:39]: added these changes to the code

File: src/main/java/org/acme/consumer/CweConsumer.java (hunk 5 above)
Comment (mehab) [reply src/main/java/org/acme/consumer/CweConsumer.java:29]: removed this one

File: src/main/java/org/acme/consumer/PrimaryConsumer.java (hunk 8 above)
Comment (mehab) [reply src/main/java/org/acme/consumer/PrimaryConsumer.java:65]: added this change :)

File: src/main/java/org/acme/consumer/OSSIndexBatcher.java (hunk 6 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/OSSIndexBatcher.java:33]: Hi Alioune....for OSS Index Batcher, we are actually calling the processor after the capacity is reached to achieve the pagination we were doing in DT and punctuator is implemented in case we dont received the mentioned capacity by the time period then we should go ahead and call the OSS API with available number of components.
As I am going through windowed aggregation that you mentioned, I see grouping only on time window, can we also do count based grouping using windowed aggregation?

Also, we are not planning to do aggregation for Snyk analyzer, that is why the batching logic is implemented in OSSIndexBatcher only and not in PrimaryConsumer.

Would look for your further input on this.

File: src/main/java/org/acme/consumer/PrimaryConsumer.java (hunk 9 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/PrimaryConsumer.java:73]: This is done

File: src/main/java/org/acme/consumer/SnykAnalyzer.java (hunk 11 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/SnykAnalyzer.java:47]: Looking into it

File: src/main/java/org/acme/Main.java (hunk 1 above)
Comment (syalioune) [reply src/main/java/org/acme/Main.java:29]: Great

File: src/main/java/org/acme/consumer/ConfigConsumer.java (hunk 4 above)
Comment (syalioune) [reply src/main/java/org/acme/consumer/ConfigConsumer.java:30]: > For this one we had a slightly modified idea around what you proposed. Let’s have a notification topic that sends a notification to our stateless quarkus app whenever there is a change in DT config. If there is a change in config then this notification topic will have an event. We want this event to be listened to by one consumer. When an event occurs, it gets the latest config from DT and updates the info in the in memory structure of quarkus apps. This way we would not be storing global config as such in ktable (like suggested yesterday) and we would only be fetching latest config from dt when there is a change (otherwise use existing values) — so there is no unnecessary polling. I think it is a good idea. What do you think?

That's an interesting alternative. There would be three things to consider here 

1. Multi-instance target : All of the instances of the quarkus stateless app should receive the event to update their in-memory state. Right now, since you are using the same `common.consumer-config-app-id` which translate to the kafka consumer `group.id`, the events will be partioned and partially processed by your cluster of quarkus app because of Kafka consumer group protocol. You need to use a random application identifier like `common.consumer-config-app-id` + `generated UUID` to have each instance receive all of the events
2. Initial state : How do you build the in-memory state at startup ? That is a challenge because even though you can manually send event for your tests, you have to take into account, instance restart or reschedule. There's a partial solution with the random application identifier I suggested above because using a combination of random `APPLICATION_ID` and `AUTO_OFFSET_RESEST=earliest` will make the consumer consume all of the events available in the topic to reconstruct it's in-memory state. That's a partial solution because you also have to take into account **Topic purge** and **Topic volumetry**. To completely solve this part, you need to make the config topic [a log compacted one](https://kafka.apache.org/documentation/#compaction)
3.  Kafka Streams usage : Kafka streams use case is read/process/write from/to Kafka. If you only need to read from Kafka to build in-memory state, there is no need for a Kafka Streams topology

I'm still concerned by this design because, at a given moment, due to consumption lag or network issues, different components from the same BOM can be processed by different instances of quarkus stateless app with different in-memory states (`snyk.enable` being simultaneously `true` and `false`). Debugging issues like that will prove very difficult

Is it really that much of an issue to read properties from DT ApiServer REST API ? At the very least, It should be benched.

File: src/main/java/org/acme/consumer/OSSIndexBatcher.java (hunk 6 above)
Comment (syalioune) [reply src/main/java/org/acme/consumer/OSSIndexBatcher.java:33]: > Hi Alioune....for OSS Index Batcher, we are actually calling the processor after the capacity is reached to achieve the pagination we were doing in DT and punctuator is implemented in case we dont received the mentioned capacity by the time period then we should go ahead and call the OSS API with available number of components. As I am going through windowed aggregation that you mentioned, I see grouping only on time window, can we also do count based grouping using windowed aggregation?

Grouping can only be done by time window. I Think that only timed aggregation could be enough. Both approaches cover the batching requirements, so way have a look at the drawback 

1. With capacity-based grouping, the processor have to know two informations : the max capacity and the time window. You need to make sure you don't get stuck into your processor if there are not enough components hence your punctuator. Because the grouping is done beforehand, you removed the pagination from `OssIndexAnalysisTask` but I think it is mostly the responsibility of each analyzer to know about its pagination method. `OssIndexAnalysisTask` should continue to paginate because you cannot enforce that Kafka Streams will always be its source and the data will be correctly paginated before reaching the analyzer ==> Responsibilities are mixed here or a tight coupling is supposed. Moreover, you cannot handle the situation where you have batched together 100 components and 98 of them are still in the cache ==> In the end, you are going to send a batch of 2 components. 
2. With time-based grouping, Kafka Streams only have to know about the time windows. You need however to handle the case where you aggregated much more components than you wanted but that's not a problem if the analyzer is self-sufficient in its batching (that's was already the case). Responsibilities are clear and you can handle the cache case, if you receive a group of 250 components with 150 still in the cache, you can filter out those still in cache before building on full batch instead of 3 incomplete batches (better rate limiting handling). You can also call the analyzer task outside the context of Kafka Streams topology and still have it perform the pagination

Also even though, you are not currently doing batch call for Snyk because the API does not expose such feature, all analyzers are designed to take a list of component (i.e `void analyze(List<Component> components)`). The component will be more flexible if you handle snyk specifics in the task rather than upstream in the Kafka Streams topology. Easier to handle a pagination of 1 than to implement pagination afterward.


To sum up my thoughts : Batching should be generic on Kafka Streams side to favor pagination regardless of downstream and pagination logic should be specific by analyzer regardless of upstream.

File: src/main/java/org/acme/consumer/ConfigConsumer.java (hunk 4 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/ConfigConsumer.java:30]: Thanks for your input Alioune..I will take into consideration the suggestions provided by you..Regarding fetching config from Dt every time, we think we would bombard DT with too many post requests as we will then have to make request for each component, as DT is sending individual component request to quarkus service

Comment (syalioune) [reply src/main/java/org/acme/consumer/ConfigConsumer.java:30]: > Regarding fetching config from Dt every time, we think we would bombard DT with too many post requests as we will then have to make request for each component

That would not be mandatory as there would be a component aggregation with Kafka Streams, you can perform the call once per invocation of `analyze(List<Component> components)` ==> One call per analyzer per window aggregation period. That could further be cached.

Comment (nscuro) [reply src/main/java/org/acme/consumer/ConfigConsumer.java:30]: I also think it's probably not as big of a problem to call the API server often, assuming some sort of caching is in place. 

> At the very least, It should be benched.

That is the key takeaway here I assume. Should be trivial to setup using something like https://github.com/rakyll/hey

File: src/main/java/org/acme/consumer/OSSIndexBatcher.java (hunk 6 above)
Comment (japurva1502) [reply src/main/java/org/acme/consumer/OSSIndexBatcher.java:33]: This is implemented

Comment (japurva1502) [reply src/main/java/org/acme/consumer/OSSIndexBatcher.java:33]: Thanks for the inputs and explanation..This has been implemented

File: src/main/java/org/acme/consumer/PrimaryConsumer.java (hunk 10 above)
Comment (syalioune) [reply src/main/java/org/acme/consumer/PrimaryConsumer.java:76]: See https://github.com/mehab/DTKafkaPOC/issues/50#issuecomment-1288799243

File: src/main/java/org/acme/consumer/CacheReader.java (hunk 3 above)
Comment (syalioune) [reply src/main/java/org/acme/consumer/CacheReader.java:57]: See https://github.com/mehab/DTKafkaPOC/issues/50#issuecomment-1288799243
